orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/Stage.java

```diff
@@ -384,81 +369,76 @@ public Task taskById(String taskId) {
-      // Get parent stages, but exclude already visited ones:
-
-      List<Stage> ancestors = new ArrayList<>();
-      if (getSyntheticStageOwner() == SyntheticStageOwner.STAGE_AFTER) {
-        ancestors.addAll(
+    toVisit.add(this);
+
+    while (!toVisit.isEmpty()) {
+      Stage curStage = toVisit.remove();
+
+      if (visited.contains(curStage.id)) {
+        continue;
+      }
+
+      visited.add(curStage.id);
+      allAncestorStages.add(curStage);
+
+      // 1. Find all stages we depend on through requisiteStageRefIds and push them on the queue
+      if (!curStage.requisiteStageRefIds.isEmpty() && !directParentOnly) {
+        List<Stage> refIdPriorStages =
             execution.getStages().stream()
-                .filter(
-                    it ->
-                        parentStageId.equals(it.parentStageId)
-                            && it.getSyntheticStageOwner() == SyntheticStageOwner.STAGE_BEFORE)
-                .collect(toList()));
+                .filter(it -> curStage.requisiteStageRefIds.contains(it.refId))
+                .collect(toList());
+
+        toVisit.addAll(refIdPriorStages);
       }
 
-      ancestors.addAll(
-          execution.getStages().stream()
-              .filter(it -> it.id.equals(parentStageId))
-              .findFirst()
-              .<List<Stage>>map(
-                  parent ->
-                      ImmutableList.<Stage>builder()
-                          .add(parent)
-                          .addAll(parent.getAncestorsImpl(visited, directParentOnly))
-                          .build())
-              .orElse(emptyList()));
-
-      return ancestors;
-    } else {
-      return emptyList();
+      // 2. If we have a parent stage
+      if (curStage.parentStageId != null && !visited.contains(curStage.parentStageId)) {
+        List<Stage> ancestors = new ArrayList<>();
+
+        // 2.1 AND we are an STAGE_AFTER, find all the stages with same parent as us but that are
+        // marked STAGE_BEFORE
+        if (curStage.getSyntheticStageOwner() == SyntheticStageOwner.STAGE_AFTER) {
+          ancestors.addAll(
+              execution.getStages().stream()
+                  .filter(
+                      it ->
+                          curStage.parentStageId.equals(it.parentStageId)
+                              && it.getSyntheticStageOwner() == SyntheticStageOwner.STAGE_BEFORE)
+                  .collect(toList()));
+        }
+
+        // 2.2 Add our direct parent to the queue
+        try {
+          ancestors.add(curStage.getParent());
+        } catch (IllegalArgumentException e) {
+          // It's not really possible for the parent not to exist.. But unittests beg to differ
+          // Not logging anything here since not having a parent would have failed and loffed in
+          // other places
+        }
+        toVisit.addAll(ancestors);
+      }
     }
+
+    return ImmutableList.<Stage>builder().addAll(allAncestorStages).build();
```

Comment:
This would be clearer as `ImmutableList.copyOf(allAncestorStages)`.  Or, alternately (and perhaps better) `allAncestorStages` could be an `ImmutableList.Builder` that we add to during the function, then the return statement would just be `return allAncestorStages.build()`.